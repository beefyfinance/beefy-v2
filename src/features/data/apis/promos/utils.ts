export function extractChainId(path: string): string {
  const match = path.match(/.+\/chain\/(.+).json$/);
  const chainId = match?.[1];
  if (chainId) {
    return chainId;
  }
  throw new Error(`Could not extract chain id from path: ${path}`);
}
