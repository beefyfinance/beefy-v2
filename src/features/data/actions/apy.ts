import { createAsyncThunk } from '@reduxjs/toolkit';
import { BeefyAPIBreakdownResponse, BeefyAPIHistoricalAPYResponse } from '../apis/beefy';
import { getBeefyApi } from '../apis/instances';

export const fetchApyAction = createAsyncThunk<BeefyAPIBreakdownResponse, {}>(
  'prices/fetchApy',
  async () => {
    const api = getBeefyApi();
    const prices = await api.getBreakdown();
    return prices;
  }
);

export const fetchHistoricalApy = createAsyncThunk<BeefyAPIHistoricalAPYResponse, {}>(
  'prices/fetchHistoricalApy',
  async () => {
    const api = getBeefyApi();
    const prices = await api.getHistoricalAPY();
    return prices;
  }
);
