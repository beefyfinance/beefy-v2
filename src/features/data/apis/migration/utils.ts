import { createAsyncThunk } from '@reduxjs/toolkit';
import type {
  CommonMigrationUpdateFulfilledPayload,
  MigratorExecuteProps,
  MigratorUpdateProps,
} from './migration-types';
import type { BeefyState } from '../../../../redux-types';
import { selectVaultById } from '../../selectors/vaults';
import { selectTokenByAddress } from '../../selectors/tokens';
import { selectUserBalanceToMigrateByVaultId } from '../../selectors/migration';
import type { Step } from '../../reducers/wallet/stepper';
import { walletActions } from '../../actions/wallet-actions';
import { startStepperWithSteps } from '../../actions/stepper';
import { isTokenErc20 } from '../../entities/token';
import { selectAllowanceByTokenAddress } from '../../selectors/allowances';
import type { VaultEntity } from '../../entities/vault';
import BigNumber from 'bignumber.js';
import type Web3 from 'web3';
import { selectChainById } from '../../selectors/chains';
import { getWalletConnectionApi, getWeb3Instance } from '../instances';

export function buildFetchBalance(
  id: string,
  fetchBalance: (
    vault: VaultEntity,
    web3: Web3,
    walletAddress: string,
    state: BeefyState
  ) => Promise<string>
) {
  return createAsyncThunk<
    CommonMigrationUpdateFulfilledPayload,
    MigratorUpdateProps,
    { state: BeefyState }
  >(`migration/${id}/update`, async ({ vaultId, walletAddress }, { getState }) => {
    const state = getState();
    const vault = selectVaultById(state, vaultId);
    const depositToken = selectTokenByAddress(state, vault.chainId, vault.depositTokenAddress);
    const chain = selectChainById(state, vault.chainId);
    const web3 = await getWeb3Instance(chain);

    const balance = await fetchBalance(vault, web3, walletAddress, state);
    const fixedBalance = new BigNumber(balance).shiftedBy(-depositToken.decimals);
    console.debug(id, vault.id, fixedBalance.toNumber());
    return { vaultId, walletAddress, balance: fixedBalance, migrationId: id };
  });
}

export function buildExecute(
  id: string,
  unstakeCall: (
    vault: VaultEntity,
    web3: Web3,
    amount: BigNumber,
    state: BeefyState
    // eslint-disable-next-line
  ) => Promise<any>
) {
  return createAsyncThunk<void, MigratorExecuteProps, { state: BeefyState }>(
    `migration/${id}/execute`,
    async ({ vaultId, t, migrationId }, { getState, dispatch }) => {
      const steps: Step[] = [];
      const state = getState();
      const vault = selectVaultById(state, vaultId);
      const depositToken = selectTokenByAddress(state, vault.chainId, vault.depositTokenAddress);
      const walletApi = await getWalletConnectionApi();
      const web3 = await walletApi.getConnectedWeb3Instance();
      const { balance } = selectUserBalanceToMigrateByVaultId(state, vaultId, migrationId);

      const call = await unstakeCall(vault, web3, balance, state);

      steps.push({
        step: 'migration',
        message: t('Vault-MigrationStart'),
        action: walletActions.migrateUnstake(
          call,
          vault,
          balance.shiftedBy(depositToken.decimals),
          migrationId
        ),
        pending: false,
        extraInfo: { vaultId },
      });

      if (isTokenErc20(depositToken)) {
        const allowance = selectAllowanceByTokenAddress(
          state,
          depositToken.chainId,
          depositToken.address,
          vault.earnContractAddress
        );
        if (allowance.lt(balance)) {
          steps.push({
            step: 'approve',
            message: t('Vault-ApproveMsg'),
            action: walletActions.approval(depositToken, vault.earnContractAddress),
            pending: false,
          });
        }
      }

      steps.push({
        step: 'deposit',
        message: t('Vault-TxnConfirm', { type: t('Deposit-noun') }),
        action: walletActions.deposit(vault, balance, true),
        pending: false,
        extraInfo: { vaultId: vault.id },
      });

      dispatch(startStepperWithSteps(steps, vault.chainId));
    }
  );
}
