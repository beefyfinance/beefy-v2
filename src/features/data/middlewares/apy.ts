import type { BeefyState } from '../../../redux-types';
import { createListenerMiddleware, isFulfilled } from '@reduxjs/toolkit';
import { reloadBalanceAndAllowanceAndGovRewardsAndBoostData } from '../actions/tokens';
import { fetchApyAction, recalculateTotalApyAction } from '../actions/apy';
import { fetchAllContractDataByChainAction } from '../actions/contract-data';

const apyListener = createListenerMiddleware<BeefyState>();

apyListener.startListening({
  matcher: isFulfilled(
    fetchApyAction,
    fetchAllContractDataByChainAction,
    reloadBalanceAndAllowanceAndGovRewardsAndBoostData
  ),
  effect: async (action, { dispatch, delay, cancelActiveListeners }) => {
    // Cancel other instances of this callback
    cancelActiveListeners();

    // Debounce
    await delay(50);

    // Compute total apy for frontend
    dispatch(recalculateTotalApyAction());
  },
});

export const apyMiddleware = apyListener.middleware;
