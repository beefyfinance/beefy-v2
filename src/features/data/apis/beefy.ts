import axios, { AxiosInstance } from 'axios';
import BigNumber from 'bignumber.js';
import { isString } from 'lodash';
import { ChainEntity } from '../entities/chain';
import { TokenEntity } from '../entities/token';
import { VaultEntity } from '../entities/vault';
import { mapValuesDeep } from '../utils/array-utils';
import { featureFlag_simulateBeefyApiError } from '../utils/feature-flags';

interface ApyGovVault {
  vaultApr: number;
}
interface ApyMaxiVault {
  totalApy: number;
}
interface ApyStandard {
  beefyPerformanceFee: number;
  vaultApr: number;
  compoundingsPerYear: number;
  vaultApy: number;
  tradingApr?: number;
  totalApy: number;
  // todo: does it make sense to have fees and apy in the same entities?
  lpFee: number;
}
export type ApyData = ApyGovVault | ApyMaxiVault | ApyStandard;
export function isStandardVaultApy(apy: ApyData): apy is ApyStandard {
  return 'compoundingsPerYear' in apy;
}
export function isGovVaultApy(apy: ApyData): apy is ApyGovVault {
  return 'vaultApr' in apy && !('compoundingsPerYear' in apy);
}
export function isMaxiVaultApy(apy: ApyData): apy is ApyMaxiVault {
  return 'totalApy' in apy && !('compoundingsPerYear' in apy);
}

export interface BeefyAPITokenPricesResponse {
  [tokenId: TokenEntity['id']]: number;
}
export interface BeefyAPIBreakdownResponse {
  [vaultId: VaultEntity['id']]: ApyData;
}

export interface BeefyAPIBuybackResponse {
  // those are of type string but they represent numbers
  [chainId: ChainEntity['id']]: { buybackTokenAmount: BigNumber; buybackUsdAmount: BigNumber };
}

// note that there is more infos but we don't need it
type BeefyAPIVaultsResponse = {
  id: string;
  lastHarvest?: number | string;
}[];

export class BeefyAPI {
  public api: AxiosInstance;
  public data: AxiosInstance;

  constructor() {
    // this could be mocked by passing mock axios to the constructor
    this.api = axios.create({
      baseURL: 'https://api.beefy.finance',
      timeout: 30 * 1000,
    });
    this.data = axios.create({
      baseURL: 'https://data.beefy.finance',
      timeout: 30 * 1000,
    });
  }

  // here we can nicely type the responses
  public async getPrices(): Promise<BeefyAPITokenPricesResponse> {
    if (featureFlag_simulateBeefyApiError('prices')) {
      throw new Error('Simulated beefy api error');
    }

    const res = await this.api.get('/prices', { params: { _: this.getCacheBuster('hour') } });
    return res.data;
  }

  // i'm not 100% certain about the return type
  // are those token ids ?
  public async getLPs(): Promise<BeefyAPITokenPricesResponse> {
    if (featureFlag_simulateBeefyApiError('lps')) {
      throw new Error('Simulated beefy api error');
    }

    const res = await this.api.get('/lps', { params: { _: this.getCacheBuster('hour') } });
    return res.data;
  }

  public async getBreakdown(): Promise<BeefyAPIBreakdownResponse> {
    if (featureFlag_simulateBeefyApiError('apy')) {
      throw new Error('Simulated beefy api error');
    }

    const res = await this.api.get('/apy/breakdown', {
      params: { _: this.getCacheBuster('day') },
    });

    // somehow, all vaultApr are currently strings, we need to fix that before sending
    // the data to be processed
    const data = mapValuesDeep(res.data, (val, key) => {
      if (key === 'vaultApr' && isString(val)) {
        val = parseFloat(val);
      }
      return val;
    });

    return data;
  }

  public async getBuyBack(): Promise<BeefyAPIBuybackResponse> {
    if (featureFlag_simulateBeefyApiError('buyback')) {
      throw new Error('Simulated beefy api error');
    }

    type ResponseType = {
      data: {
        [chainId: ChainEntity['id']]: {
          buybackTokenAmount: string;
          buybackUsdAmount: string;
        };
      };
    };
    const strRes = await this.api.get<ResponseType>('/bifibuyback', {
      params: { _: this.getCacheBuster('hour') },
    });
    // format result with big number as we get strings from the api
    const res: BeefyAPIBuybackResponse = {};
    for (const chainId in strRes.data) {
      res[chainId] = {
        buybackTokenAmount: new BigNumber(strRes.data[chainId].buybackTokenAmount),
        buybackUsdAmount: new BigNumber(strRes.data[chainId].buybackUsdAmount),
      };
    }
    return res;
  }

  /**
   * For now we fetch lastHarvest from the api
   * TODO: fetch this from the contract directly
   */
  public async getVaultLastHarvest(vaultId: VaultEntity['id']): Promise<null | Date> {
    const res = await this.api.get<BeefyAPIVaultsResponse>('/vaults', {
      params: { _: this.getCacheBuster('hour') },
    });

    // const vault = data.filter(vault => vault.id.includes(vaultId));
    const vaultConfig = res.data.find(vault => {
      return vault.id === vaultId;
    });

    if (!vaultConfig) {
      // vault is not harvestable (most probably a gov vault)
      return null;
    }
    if (!('lastHarvest' in vaultConfig)) {
      return null;
    }
    let lh = 0;
    if (isString(vaultConfig.lastHarvest)) {
      lh = parseInt(vaultConfig.lastHarvest);
    } else {
      lh = vaultConfig.lastHarvest;
    }
    if (lh === 0) {
      return null;
    }

    return new Date(lh * 1000);
  }

  // maybe have a local cache instead of this cache busting
  // have to check if this returns browser cache before doing so
  protected getCacheBuster(mode: 'hour' | 'day' = 'day'): number {
    if (mode === 'hour') {
      return Math.trunc(Date.now() / (1000 * 60));
    } else {
      const cache = new Date();
      cache.setMinutes(0, 0, 0);
      return cache.getTime();
    }
  }
}
