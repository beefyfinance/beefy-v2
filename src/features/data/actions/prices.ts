import { createAsyncThunk } from '@reduxjs/toolkit';
import { BeefyAPITokenPricesResponse } from '../apis/beefy';
import { getBeefyApi } from '../apis/instances';

export const fetchAllPricesAction = createAsyncThunk<BeefyAPITokenPricesResponse, {}>(
  'prices/fetchAllPricesAction',
  async () => {
    const api = getBeefyApi();
    const [prices, lpPrices] = await Promise.all([api.getPrices(), api.getLPs()]);
    return {
      ...prices,
      ...lpPrices,
    };
  }
);
