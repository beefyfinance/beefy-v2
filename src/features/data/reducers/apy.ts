import { createSlice } from '@reduxjs/toolkit';
import { WritableDraft } from 'immer/dist/internal';
import { BeefyState } from '../../../redux-types';
import { fetchApyAction } from '../actions/apy';
import { fetchAllContractDataByChainAction } from '../actions/contract-data';
import { reloadBalanceAndAllowanceAndGovRewardsAndBoostData } from '../actions/tokens';
import { ApyData } from '../apis/beefy';
import {
  BoostContractData,
  FetchAllContractDataResult,
} from '../apis/contract-data/contract-data-types';
import { BoostEntity } from '../entities/boost';
import { isGovVault, VaultEntity } from '../entities/vault';
import {
  selectActiveVaultBoostIds,
  selectBoostById,
  selectIsVaultBoosted,
} from '../selectors/boosts';
import { selectIsConfigAvailable } from '../selectors/data-loader';
import { selectTokenByAddress, selectTokenPriceByAddress } from '../selectors/tokens';
import { selectVaultById, selectVaultPricePerFullShare } from '../selectors/vaults';
import { createIdMap } from '../utils/array-utils';
import { mooAmountToOracleAmount } from '../utils/ppfs';
import { getBoostStatusFromPeriodFinish } from './boosts';
import { BIG_ONE } from '../../../helpers/big-number';

// boost is expressed as APR
interface AprData {
  apr: number;
}

// TODO: this should be reworked
export interface TotalApy {
  totalApy?: number;
  totalDaily?: number;
  vaultApr?: number;
  vaultDaily?: number;
  tradingApr?: number;
  tradingDaily?: number;
  boostApr?: number;
  boostDaily?: number;
  boostedTotalApy?: number;
  boostedTotalDaily?: number;
}

/**
 * State containing APY infos indexed by vault id
 */
export interface ApyState {
  rawApy: {
    byVaultId: {
      // we reuse the api types, not the best idea but works for now
      [vaultId: VaultEntity['id']]: ApyData;
    };
    byBoostId: {
      [boostId: BoostEntity['id']]: AprData;
    };
  };
  totalApy: {
    byVaultId: {
      // we reuse the api types, not the best idea but works for now
      [vaultId: VaultEntity['id']]: TotalApy;
    };
  };
}
export const initialApyState: ApyState = {
  rawApy: { byVaultId: {}, byBoostId: {} },
  totalApy: { byVaultId: {} },
};

export const apySlice = createSlice({
  name: 'apy',
  initialState: initialApyState,
  reducers: {
    // standard reducer logic, with auto-generated action types per reducer
  },
  extraReducers: builder => {
    builder.addCase(fetchApyAction.fulfilled, (sliceState, action) => {
      for (const [vaultId, apy] of Object.entries(action.payload.data)) {
        sliceState.rawApy.byVaultId[vaultId] = apy;
      }

      // recompute total apy to have it ready to render vault list super fast
      const state = action.payload.state;
      const updatedState: BeefyState = { ...state, biz: { ...state.biz, apy: sliceState } };
      recomputeTotalApy(updatedState, sliceState);
    });

    builder.addCase(fetchAllContractDataByChainAction.fulfilled, (sliceState, action) => {
      const state = action.payload.state;
      addContractDataToState(state, sliceState, action.payload.data);
    });

    builder.addCase(
      reloadBalanceAndAllowanceAndGovRewardsAndBoostData.fulfilled,
      (sliceState, action) => {
        const state = action.payload.state;
        addContractDataToState(state, sliceState, action.payload.contractData);
      }
    );
  },
});

function addContractDataToState(
  state: BeefyState,
  sliceState: WritableDraft<ApyState>,
  contractData: FetchAllContractDataResult
) {
  const now = new Date();
  const activeBoostsByVaultIds: { [vaultId: VaultEntity['id']]: BoostContractData[] } = {};

  // create a quick access map
  const vaultDataByVaultId = createIdMap(contractData.standardVaults);

  for (const boostContractData of contractData.boosts) {
    const boost = selectBoostById(state, boostContractData.id);
    const vault = selectVaultById(state, boost.vaultId);
    const boostStatus = getBoostStatusFromPeriodFinish(boostContractData.periodFinish, now);
    // boost is expired, don't count apy
    if (boostStatus === 'expired') {
      continue;
    }
    if (activeBoostsByVaultIds[vault.id] === undefined) {
      activeBoostsByVaultIds[vault.id] = [];
    }
    activeBoostsByVaultIds[vault.id].push(boostContractData);

    /**
     * Some boosts can yield rewards in mooToken, be it from the same vault
     * or for another vault (like binSPIRIT) and some boost yield rewards as another
     * unrelated token (like PAE). When the boost is a mooToken, we don't have the
     * token price in the api so we need to compute it.
     **/
    let earnedPrice = null;
    const rewardTargetVaultId =
      state.entities.vaults.byChainId[boost.chainId]?.standardVault.byEarnedTokenAddress[
        boost.earnedTokenAddress.toLowerCase()
      ];
    if (rewardTargetVaultId) {
      const rewardTargetVault = selectVaultById(state, rewardTargetVaultId);
      const rewardVaultOraclePrice = selectTokenPriceByAddress(
        state,
        boost.chainId,
        rewardTargetVault.depositTokenAddress
      );
      const depositToken = selectTokenByAddress(
        state,
        boost.chainId,
        rewardTargetVault.depositTokenAddress
      );
      const earnedToken = selectTokenByAddress(state, boost.chainId, boost.earnedTokenAddress);

      // use the latest ppfs if any, otherwise fetch it in the previous state
      const ppfs =
        vaultDataByVaultId[rewardTargetVault.id]?.pricePerFullShare ||
        selectVaultPricePerFullShare(state, rewardTargetVault.id);

      // so the price rate is the oracle token price by the rate of the conversion to mooToken
      const mooToOracleRate = mooAmountToOracleAmount(earnedToken, depositToken, ppfs, BIG_ONE);
      earnedPrice = mooToOracleRate.times(rewardVaultOraclePrice);
    } else {
      // if we don't have a matching vault, it should be a yield from a token that has a price
      earnedPrice = selectTokenPriceByAddress(state, boost.chainId, boost.earnedTokenAddress);
    }

    const stakedTokenPrice = selectTokenPriceByAddress(
      state,
      vault.chainId,
      vault.depositTokenAddress
    );

    const ppfs =
      vaultDataByVaultId[boost.vaultId]?.pricePerFullShare ||
      selectVaultPricePerFullShare(state, boost.vaultId);
    const totalStakedInUsd = boostContractData.totalSupply.times(stakedTokenPrice).times(ppfs);
    const yearlyRewardsInUsd = boostContractData.rewardRate
      .times(3600 * 24 * 365)
      .times(earnedPrice);

    const apr = yearlyRewardsInUsd.dividedBy(totalStakedInUsd).toNumber();

    // add data to state
    sliceState.rawApy.byBoostId[boost.id] = { apr };
  }

  // recompute total apy to have it ready to render vault list super fast
  const updatedState: BeefyState = { ...state, biz: { ...state.biz, apy: sliceState } };
  recomputeTotalApy(updatedState, sliceState, activeBoostsByVaultIds);
}

function recomputeTotalApy(
  state: BeefyState,
  sliceState: WritableDraft<ApyState>,
  // we need up to date boost data to know if we need to include boost apy
  newActiveBoostsByVaultIds: { [vaultId: VaultEntity['id']]: BoostContractData[] } = {}
) {
  // no point in computing if all data is not loaded
  if (!selectIsConfigAvailable(state)) {
    return;
  }

  for (const [vaultId, apyData] of Object.entries(sliceState.rawApy.byVaultId)) {
    const values: TotalApy = {};

    // sometimes we get vault ids in the api that are not yet configure
    // locally, so we have to check that the vault exists first
    if (state.entities.vaults.byId[vaultId] === undefined) {
      continue;
    }

    const vault = selectVaultById(state, vaultId);
    const isBoosted =
      newActiveBoostsByVaultIds[vaultId] !== undefined || selectIsVaultBoosted(state, vaultId);

    let boostApr = 0;
    if (isBoosted) {
      // todo: not sure why but we only use the first
      const latestActiveBoostId =
        newActiveBoostsByVaultIds[vaultId] !== undefined
          ? newActiveBoostsByVaultIds[vaultId][0].id
          : selectActiveVaultBoostIds(state, vaultId)[0];
      boostApr = sliceState.rawApy.byBoostId[latestActiveBoostId]?.apr || 0;
    }
    const apy = apyData as any; /* Legacy code fix */

    values.totalApy = apy.totalApy;

    if (apy.vaultApr) {
      values.vaultApr = apy.vaultApr;
      values.vaultDaily = apy.vaultApr / 365;
    }

    if (apy.tradingApr) {
      values.tradingApr = apy.tradingApr;
      values.tradingDaily = apy.tradingApr / 365;
    }

    if (values.vaultDaily || values.tradingDaily) {
      values.totalDaily = (values.vaultDaily || 0) + (values.tradingDaily || 0);
    } else {
      values.totalDaily = yearlyToDaily(values.totalApy);
    }

    if (isGovVault(vault)) {
      values.totalApy = apy.vaultApr / 1;
      values.totalDaily = apy.vaultApr / 365;
    }

    if (isBoosted) {
      values.boostApr = boostApr;
      values.boostDaily = boostApr / 365;
      values.boostedTotalApy = values.boostApr ? values.totalApy + values.boostApr : 0;
      values.boostedTotalDaily = values.boostDaily ? values.totalDaily + values.boostDaily : 0;
    }

    sliceState.totalApy.byVaultId[vaultId] = values;
  }
}

const yearlyToDaily = (apy: number) => {
  const g = Math.pow(10, Math.log10(apy + 1) / 365) - 1;

  if (isNaN(g)) {
    return 0;
  }

  return g;
};
