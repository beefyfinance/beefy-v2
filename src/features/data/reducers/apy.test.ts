import BigNumber from 'bignumber.js';
import { FetchAllApyFulfilledPayload, fetchApyAction } from '../actions/apy';
import {
  fetchAllContractDataByChainAction,
  FetchAllContractDataFulfilledPayload,
} from '../actions/contract-data';
import { fetchAllPricesAction } from '../actions/prices';
import { BeefyAPITokenPricesResponse } from '../apis/beefy';
import { getBeefyTestingStore } from '../utils/test-utils';
import { apySlice, initialApyState } from './apy';

describe('APY slice tests', () => {
  it('should update state on fulfilled apy data action', async () => {
    // we have loaded some entities already
    const store = await getBeefyTestingStore();
    const state = store.getState();

    // We want to make sure we handle the action properly
    const payload: FetchAllApyFulfilledPayload = {
      data: {
        // one standard vault
        'ellipsis-renbtc': {
          vaultApr: 0.03346493761979882,
          compoundingsPerYear: 2190,
          beefyPerformanceFee: 0.045,
          vaultApy: 0.03403092311211742,
          lpFee: 0.0002,
          tradingApr: 0.000309155,
          totalApy: 0.0343505989421522,
        },
        // one boosted standard vault
        'banana-banana-busd': {
          vaultApr: 0.03346493761979882,
          compoundingsPerYear: 2190,
          beefyPerformanceFee: 0.045,
          vaultApy: 0.03403092311211742,
          lpFee: 0.0002,
          tradingApr: 0.000309155,
          totalApy: 0.0343505989421522,
        },
        // one gov vault
        'one-bifi-gov': { vaultApr: 0.01385586450952859762 },
        // one maxi vault
        'one-bifi-maxi': { totalApy: 0.01401646541942525 },

        // one that does not exists locally yet
        'future-extra-super-hyped-vault': { vaultApr: 0.01385586450952859762 },
      },
      state,
    };
    const action = { type: fetchApyAction.fulfilled, payload: payload };
    const newState = apySlice.reducer(initialApyState, action);
    expect(newState).toMatchSnapshot();

    // now, we get a boost info
    // we should compute total boost apr/y
    const boostPayload: FetchAllContractDataFulfilledPayload = {
      chainId: 'bsc',
      data: {
        govVaults: [],
        standardVaults: [],
        boosts: [
          {
            id: 'moo_banana-banana-busd-bitcrush',
            periodFinish: new Date(2050, 0, 1, 0, 0, 0),
            rewardRate: new BigNumber('123'),
            totalSupply: new BigNumber('12345'),
          },
        ],
      },
      state,
    };
    const boostAction = {
      type: fetchAllContractDataByChainAction.fulfilled,
      payload: boostPayload,
    };
    const newBoostedState = apySlice.reducer(newState, boostAction);
    expect(newBoostedState).toMatchSnapshot();
  });

  it('should update state on fulfilled boost contract data', async () => {
    // we have loaded some entities already
    const store = await getBeefyTestingStore();
    const state = store.getState();

    // We want to make sure we handle the action properly
    const payload: FetchAllContractDataFulfilledPayload = {
      chainId: 'bsc',
      data: {
        govVaults: [],
        standardVaults: [],
        boosts: [
          {
            id: 'moo_banana-banana-busd-bitcrush',
            periodFinish: new Date(2050, 0, 1, 0, 0, 0),
            rewardRate: new BigNumber(0.4),
            totalSupply: new BigNumber(12345),
          },
        ],
      },
      state,
    };
    const action = { type: fetchAllContractDataByChainAction.fulfilled, payload: payload };
    const newState = apySlice.reducer(initialApyState, action);
    expect(newState).toMatchSnapshot();
  });

  it('should properly compute binSPIRIT boost apr when getting the ppfs in the same action', async () => {
    // we have loaded some entities already
    const store = await getBeefyTestingStore();

    const initPayload: BeefyAPITokenPricesResponse = {
      'spirit-binspirit-spirit': 0.2597297311519736,
      binSPIRIT: 0.10508944967219624,
    };
    store.dispatch({ type: fetchAllPricesAction.fulfilled, payload: initPayload });

    const state = store.getState();

    // We want to make sure we handle the action properly
    const payload: FetchAllContractDataFulfilledPayload = {
      chainId: 'fantom',
      data: {
        govVaults: [],
        standardVaults: [
          {
            id: 'beefy-binspirit',
            balance: new BigNumber(0),
            pricePerFullShare: new BigNumber('1.120250293767506234'),
            strategy: '0x000',
          },
        ],
        boosts: [
          {
            id: 'moo_spirit-binspirit-spirit-spirit',
            periodFinish: new Date(2050, 0, 1, 0, 0, 0),
            rewardRate: new BigNumber('0.142541908840469245'),
            totalSupply: new BigNumber('1338329.00751070894030697'),
          },
        ],
      },
      state,
    };
    const action = { type: fetchAllContractDataByChainAction.fulfilled, payload: payload };
    const newState = apySlice.reducer(initialApyState, action);
    expect(newState.rawApy.byBoostId['moo_spirit-binspirit-spirit-spirit'].apr).toMatchSnapshot();
  });

  it('should properly compute binSPIRIT boost apr when getting the ppfs beforehand', async () => {
    // we have loaded some entities already
    const store = await getBeefyTestingStore();

    const initPayload1: BeefyAPITokenPricesResponse = {
      'spirit-binspirit-spirit': 0.2597297311519736,
      binSPIRIT: 0.10508944967219624,
    };
    store.dispatch({ type: fetchAllPricesAction.fulfilled, payload: initPayload1 });
    // we get the ppfs beforehand
    const initPayload2: FetchAllContractDataFulfilledPayload = {
      chainId: 'fantom',
      data: {
        govVaults: [],
        standardVaults: [
          {
            id: 'beefy-binspirit',
            balance: new BigNumber(0),
            pricePerFullShare: new BigNumber('1.120250293767506234'),
            strategy: '0x000',
          },
        ],
        boosts: [],
      },
      state: store.getState(),
    };
    store.dispatch({ type: fetchAllContractDataByChainAction.fulfilled, payload: initPayload2 });

    const state = store.getState();

    // We want to make sure we handle the action properly
    const payload: FetchAllContractDataFulfilledPayload = {
      chainId: 'fantom',
      data: {
        govVaults: [],
        standardVaults: [],
        boosts: [
          {
            id: 'moo_spirit-binspirit-spirit-spirit',
            periodFinish: new Date(2050, 0, 1, 0, 0, 0),
            rewardRate: new BigNumber('0.142541908840469245'),
            totalSupply: new BigNumber('1338329.00751070894030697'),
          },
        ],
      },
      state,
    };
    const action = { type: fetchAllContractDataByChainAction.fulfilled, payload: payload };
    const newState = apySlice.reducer(initialApyState, action);
    expect(newState.rawApy.byBoostId['moo_spirit-binspirit-spirit-spirit'].apr).toMatchSnapshot();
  });
});
