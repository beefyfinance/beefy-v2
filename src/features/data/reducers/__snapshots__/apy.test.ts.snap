// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`APY slice tests should properly compute binSPIRIT boost apr when getting the ppfs beforehand 1`] = `1.5224351568892651`;

exports[`APY slice tests should properly compute binSPIRIT boost apr when getting the ppfs in the same action 1`] = `1.5224351568892651`;

exports[`APY slice tests should update state on fulfilled apy data action 1`] = `
Object {
  "rawApy": Object {
    "byBoostId": Object {},
    "byVaultId": Object {
      "banana-banana-busd": Object {
        "beefyPerformanceFee": 0.045,
        "compoundingsPerYear": 2190,
        "lpFee": 0.0002,
        "totalApy": 0.0343505989421522,
        "tradingApr": 0.000309155,
        "vaultApr": 0.03346493761979882,
        "vaultApy": 0.03403092311211742,
      },
      "ellipsis-renbtc": Object {
        "beefyPerformanceFee": 0.045,
        "compoundingsPerYear": 2190,
        "lpFee": 0.0002,
        "totalApy": 0.0343505989421522,
        "tradingApr": 0.000309155,
        "vaultApr": 0.03346493761979882,
        "vaultApy": 0.03403092311211742,
      },
      "future-extra-super-hyped-vault": Object {
        "vaultApr": 0.013855864509528597,
      },
      "one-bifi-gov": Object {
        "vaultApr": 0.013855864509528597,
      },
      "one-bifi-maxi": Object {
        "totalApy": 0.01401646541942525,
      },
    },
  },
  "totalApy": Object {
    "byVaultId": Object {
      "banana-banana-busd": Object {
        "totalApy": 0.0343505989421522,
        "totalDaily": 0.00009253176060218854,
        "tradingApr": 0.000309155,
        "tradingDaily": 8.469999999999999e-7,
        "vaultApr": 0.03346493761979882,
        "vaultDaily": 0.00009168476060218854,
      },
      "ellipsis-renbtc": Object {
        "totalApy": 0.0343505989421522,
        "totalDaily": 0.00009253176060218854,
        "tradingApr": 0.000309155,
        "tradingDaily": 8.469999999999999e-7,
        "vaultApr": 0.03346493761979882,
        "vaultDaily": 0.00009168476060218854,
      },
      "one-bifi-gov": Object {
        "totalApy": 0.013855864509528597,
        "totalDaily": 0.000037961272628845475,
        "vaultApr": 0.013855864509528597,
        "vaultDaily": 0.000037961272628845475,
      },
      "one-bifi-maxi": Object {
        "totalApy": 0.01401646541942525,
        "totalDaily": 0.00003813536582875088,
      },
    },
  },
}
`;

exports[`APY slice tests should update state on fulfilled apy data action 2`] = `
Object {
  "rawApy": Object {
    "byBoostId": Object {
      "moo_banana-banana-busd-bitcrush": Object {
        "apr": 4950.645066260906,
      },
    },
    "byVaultId": Object {
      "banana-banana-busd": Object {
        "beefyPerformanceFee": 0.045,
        "compoundingsPerYear": 2190,
        "lpFee": 0.0002,
        "totalApy": 0.0343505989421522,
        "tradingApr": 0.000309155,
        "vaultApr": 0.03346493761979882,
        "vaultApy": 0.03403092311211742,
      },
      "ellipsis-renbtc": Object {
        "beefyPerformanceFee": 0.045,
        "compoundingsPerYear": 2190,
        "lpFee": 0.0002,
        "totalApy": 0.0343505989421522,
        "tradingApr": 0.000309155,
        "vaultApr": 0.03346493761979882,
        "vaultApy": 0.03403092311211742,
      },
      "future-extra-super-hyped-vault": Object {
        "vaultApr": 0.013855864509528597,
      },
      "one-bifi-gov": Object {
        "vaultApr": 0.013855864509528597,
      },
      "one-bifi-maxi": Object {
        "totalApy": 0.01401646541942525,
      },
    },
  },
  "totalApy": Object {
    "byVaultId": Object {
      "banana-banana-busd": Object {
        "boostApr": 4950.645066260906,
        "boostDaily": 13.563411140440838,
        "boostedTotalApy": 4950.679416859848,
        "boostedTotalDaily": 13.56350367220144,
        "totalApy": 0.0343505989421522,
        "totalDaily": 0.00009253176060218854,
        "tradingApr": 0.000309155,
        "tradingDaily": 8.469999999999999e-7,
        "vaultApr": 0.03346493761979882,
        "vaultDaily": 0.00009168476060218854,
      },
      "ellipsis-renbtc": Object {
        "totalApy": 0.0343505989421522,
        "totalDaily": 0.00009253176060218854,
        "tradingApr": 0.000309155,
        "tradingDaily": 8.469999999999999e-7,
        "vaultApr": 0.03346493761979882,
        "vaultDaily": 0.00009168476060218854,
      },
      "one-bifi-gov": Object {
        "totalApy": 0.013855864509528597,
        "totalDaily": 0.000037961272628845475,
        "vaultApr": 0.013855864509528597,
        "vaultDaily": 0.000037961272628845475,
      },
      "one-bifi-maxi": Object {
        "totalApy": 0.01401646541942525,
        "totalDaily": 0.00003813536582875088,
      },
    },
  },
}
`;

exports[`APY slice tests should update state on fulfilled boost contract data 1`] = `
Object {
  "rawApy": Object {
    "byBoostId": Object {
      "moo_banana-banana-busd-bitcrush": Object {
        "apr": 16.099658752067988,
      },
    },
    "byVaultId": Object {},
  },
  "totalApy": Object {
    "byVaultId": Object {},
  },
}
`;
